pkg/machine/compression/gzip.go

```diff
@@ -0,0 +1,37 @@
+package compression
+
+import (
+	"io"
+
+	image "github.com/containers/image/v5/pkg/compression"
+	"github.com/sirupsen/logrus"
+)
+
+type gzipDecompressor struct {
+	genericDecompressor
+	gzReader io.ReadCloser
+}
+
+func newGzipDecompressor(compressedFilePath string) (*gzipDecompressor, error) {
+	d, err := newGenericDecompressor(compressedFilePath)
+	return &gzipDecompressor{*d, nil}, err
+}
+
+func (d *gzipDecompressor) decompress(w WriteSeekCloser, r io.Reader) error {
+	gzReader, err := image.GzipDecompressor(r)
+	if err != nil {
+		return err
+	}
+	d.gzReader = gzReader
+
+	sparseWriter := NewSparseWriter(w)
+	_, err = io.Copy(sparseWriter, gzReader)
```

Comment:
In every instance:

This never calls `sparseWrite.Close()`, and therefore corrupts data (does not write the trailing zero block, if any).

It might be simpler (and more symmetric) to have `w` just be an `io.WriteSeeker`, and to wrap it with (a functional equivalent of) `NopCloser` for the purpose of this function. … and because that repeats 3 times, perhaps do that in a local `copySparse` utility.

Reply:
(Also, again because `sparseWriter.Close` is non-trivial, its return value should be checked, not ignored.)